src/request.rs

```diff
@@ -125,16 +128,38 @@ impl Request {
             }
         };
 
-        let reader = payload.into_read();
-        let unit = Unit::new(
-            &self.agent,
-            &self.method,
-            &url,
-            self.headers,
-            &reader,
-            deadline,
-        );
-        let response = unit::connect(unit, true, reader).map_err(|e| e.url(url.clone()))?;
+        let request_fn = |req: Request| {
+            let reader = payload.into_read();
+            let unit = Unit::new(
+                &req.agent,
+                &req.method,
+                &url,
+                req.headers,
+                &reader,
+                deadline,
+            );
+
+            unit::connect(unit, true, reader).map_err(|e| e.url(url.clone()))
+        };
+
+        // This clone is quite cheap since either we are cloning the Optional::None or a Vec<Arc<dyn Middleware>>.
+        let maybe_middleware = agent.state.middleware.clone();
```

Comment:
Can MiddlewareNext instead store an iterator over the `middleware` Vec?

Reply:
Maybe. I'll take a stab at it.